### PR TITLE
PICARD-1494: Improve file times preservation using os.utime() ns parameter 

### PR DESCRIPTION
**Pull request by zas**
# Summary

* This is a…
    * [ ] Bug fix
    * [ ] Feature addition
    * [x] Refactoring
    * [ ] Minor / simple change (like a typo)
    * [ ] Other
* **Describe this change in 1-2 sentences**:

# Problem

* JIRA ticket (_optional_): [PICARD-1494](https://tickets.metabrainz.org/browse/PICARD-1494)

`os.stat()` `st_atime` and `st_mtime` have low resolution, so time preservation doesn't work that well, depending on platform (as bad as 2 seconds delta on windows).

# Solution

 Python3.3 introduces nano seconds resolution times that can be used with utimes() using `ns` parameter

# Action

### CODE REVIEW

**Comment by zas** [comment]
Perhaps fix https://tickets.metabrainz.org/browse/PICARD-1110 but it has to be tested

File: picard/file.py [hunk 1]
```diff
@@ -215,13 +242,18 @@ def _save_and_rename(self, old_filename, metadata):
...
             if config.setting["preserve_timestamps"]:
                 try:
-                    os.utime(old_filename, (info.st_atime, info.st_mtime))
-                except OSError:
-                    log.warning("Couldn't preserve timestamp for %r", old_filename)
+                    self._preserve_times(old_filename, save)
+                except self.PreserveTimesStatError as why:
+                    log.warning(why)
+                    # we didn't save the file yet, bail out
+                    return None
```
[the remark this answers is not in the document]
Comment (zas) [reply picard/file.py]: I don't think so, this error occurs when we can't stat the original file, and it usually means we'll not be able to read or write it during save phase.
It sticks to previous behavior, and IMHO makes sense to just fail when this happens.

**Comment by zas** [comment]
> But I think we should add a ticket for this. More precision in preserving timestamps is a minor improvement but IMHO still worth to being noted in the changelog.

Ticket created as https://tickets.metabrainz.org/browse/PICARD-1494

File: picard/file.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (phw) [reply picard/file.py]: Ok, yes. Agreed, it is more than unlikely any saving will succeed after this error.

**Comment by zas** [comment]
@phw 
A build job failed, but i can't really explain what happened: https://travis-ci.org/metabrainz/picard/jobs/508093204

Only one job failed over 19.

Env:
```
 Python: 3.6
INSTALL_DEPS="$PIP_INSTALL pyqt5==5.8 mutagen==1.37 discid"
```

Differences between expect and result values:
```
1552944516688544815-1552944516692544815
-4000000
1552944516692544815-1552944516688544815
4000000
```

Build log:

```
=================================== FAILURES ===================================
____________________ TestPreserveTimes.test_preserve_times _____________________
self = <test.test_file.TestPreserveTimes testMethod=test_preserve_times>
    def test_preserve_times(self):
        self._create_testfile()
    
        st = os.stat(self.file.filename)
        expect = (st.st_mtime_ns, st.st_atime_ns)
    
        # test if times are preserved
        result = self.file._preserve_times(self.file.filename,
                                           self._modify_testfile)
>       self.assertEqual(result, expect)
E       AssertionError: Tuples differ: (1552944516688544815, 1552944516692544815) != (1552944516692544815, 1552944516688544815)
E       
E       First differing element 0:
E       1552944516688544815
E       1552944516692544815
E       
E       - (1552944516688544815, 1552944516692544815)
E       + (1552944516692544815, 1552944516688544815)
test/test_file.py:69: AssertionError
```

**Comment by zas** [comment]
It happened again, in the exact same env...
I wonder if it is only an issue with the test, i'll try to fsync() test files after they are modified to see if it still happens.

**Pull request merged**